x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlan.java

```diff
@@ -0,0 +1,301 @@
+                List<Tuple<Model, Integer>> allocationsPerModel = nodeToModel.containsKey(n) ? nodeToModel.get(n) : new ArrayList<>();
+                allocationsPerModel.add(Tuple.tuple(m, assignments.get(m).get(n)));
+                nodeToModel.put(n, allocationsPerModel);
+            }
+        }
+
+        StringBuilder msg = new StringBuilder();
+        List<Node> nodes = nodeToModel.keySet().stream().sorted(Comparator.comparing(Node::id)).collect(Collectors.toList());
+        for (int i = 0; i < nodes.size(); i++) {
+            Node n = nodes.get(i);
+            msg.append(n);
+            msg.append(" ->");
+            for (Tuple<Model, Integer> modelAllocations : nodeToModel.get(n)
+                .stream()
+                .sorted(Comparator.comparing(x -> x.v1().id()))
+                .collect(Collectors.toList())) {
+                if (modelAllocations.v2() > 0) {
+                    msg.append(" ");
+                    msg.append(modelAllocations.v1().id());
+                    msg.append(" (mem = ");
+                    msg.append(ByteSizeValue.ofBytes(modelAllocations.v1().memoryBytes()));
+                    msg.append(")");
+                    msg.append(" (allocations = ");
+                    msg.append(modelAllocations.v2());
+                    msg.append("/");
+                    msg.append(modelAllocations.v1().allocations());
+                    msg.append(")");
+                    msg.append(" (threads_per_allocation = ");
+                    msg.append(modelAllocations.v1().threadsPerAllocation());
+                    msg.append(")");
+                }
+            }
+            if (i < nodes.size() - 1) {
+                msg.append('\n');
+            }
+        }
+        return msg.toString();
+    }
+
+    public static Builder builder(Collection<Node> nodes, Collection<Model> models) {
+        return new Builder(nodes, models);
+    }
+
+    static class Builder {
+
+        private final Map<Model, Map<Node, Integer>> assignments;
+        private final Map<Node, Long> remainingNodeMemory;
+        private final Map<Node, Integer> remainingNodeCores;
+        private final Map<Model, Integer> remainingModelAllocations;
+
+        private Builder(Collection<Node> nodes, Collection<Model> models) {
+            if (nodes.stream().collect(Collectors.toSet()).size() != nodes.size()) {
+                throw new IllegalArgumentException("there should be no duplicate nodes");
+            }
+            if (models.stream().collect(Collectors.toSet()).size() != models.size()) {
+                throw new IllegalArgumentException("there should be no duplicate models");
+            }
+
+            assignments = new HashMap<>(nodes.size() * models.size(), 1.0f);
+            remainingNodeMemory = new HashMap<>(nodes.size(), 1.0f);
+            remainingNodeCores = new HashMap<>(nodes.size(), 1.0f);
+            remainingModelAllocations = new HashMap<>(models.size(), 1.0f);
```

Comment:
All of these should use `Maps.
```suggestion
            assignments = Maps.newMapWithExpectedSize(nodes.size() * models.size());
            remainingNodeMemory = Maps.newMapWithExpectedSize(nodes.size());
            remainingNodeCores = Maps.newMapWithExpectedSize(nodes.size());
            remainingModelAllocations = Maps.newMapWithExpectedSize(models.size());
```